src/hotspot/share/oops/compressedKlass.cpp

```diff
@@ -96,7 +96,7 @@ void CompressedKlassPointers::sanity_check_after_initialization() {
 
   // Check that Klass range is fully engulfed in the encoding range
   const address encoding_start = _base;
-  const address encoding_end = _base + nth_bit(narrow_klass_pointer_bits() + _shift);
+  const address encoding_end = (address)(p2u(_base) + (uintptr_t)nth_bit(narrow_klass_pointer_bits() + _shift));
```

Comment:
nth_bit should already give us a 64-bit value, why the second cast?
I see that nth_bit returns an intptr_t - is the sign the problem? We may want to change that to uintptr_t...

Reply:
Prefer using `uintptr_t` because `intptr_t` has undefined behavior on overflow. Probably not in this case, here.

Reply:
It should work without the second cast, but there may be compiler warnings like "warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]".

Reply:
> It should work without the second cast, but there may be compiler warnings like "warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]".

Yes, below example can reproduce the gcc/clang compile warning.

```c
#include <stdio.h>

int main()
{
  unsigned long a = 1;
  signed long b = 2;
  printf("%llu\n", (long long unsigned int)(a + b));
  return 0;
}
```